cluster-autoscaler/cloudprovider/openstack/openstack_cloud_provider.go

```diff
@@ -0,0 +1,177 @@
+func (os *openstackCloudProvider) AddNodeGroup(group OpenstackNodeGroup) {
+	os.nodeGroups = append(os.nodeGroups, group)
+}
+
+// NodeGroupForNode returns the node group that a given node belongs to.
+//
+// Since only a single node group is currently supported, the first node group is always returned.
+func (os *openstackCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	// TODO: wait for magnum nodegroup support
+	return &(os.nodeGroups[0]), nil
+}
+
+// Pricing is not implemented.
+func (os *openstackCloudProvider) Pricing() (cloudprovider.PricingModel, errors.AutoscalerError) {
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+// GetAvailableMachineTypes is not implemented.
+func (os *openstackCloudProvider) GetAvailableMachineTypes() ([]string, error) {
+	return []string{}, nil
+}
+
+// NewNodeGroup is not implemented.
+func (os *openstackCloudProvider) NewNodeGroup(machineType string, labels map[string]string, systemLabels map[string]string,
+	taints []apiv1.Taint, extraResources map[string]resource.Quantity) (cloudprovider.NodeGroup, error) {
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+// GetResourceLimiter returns resource constraints for the cloud provider
+func (os *openstackCloudProvider) GetResourceLimiter() (*cloudprovider.ResourceLimiter, error) {
+	return os.resourceLimiter, nil
+}
+
+// Refresh is called before every autoscaler main loop.
+//
+// Currently only prints debug information.
+func (os *openstackCloudProvider) Refresh() error {
+	for _, nodegroup := range os.nodeGroups {
+		klog.V(3).Info(nodegroup.Debug())
+	}
+	return nil
+}
+
+// Cleanup currently does nothing.
+func (os *openstackCloudProvider) Cleanup() error {
+	return nil
+}
+
+// BuildOpenstack is called by the autoscaler to build an openstack cloud provider.
+//
+// The OpenstackManager is created here, and the node groups are created
+// based on the specs provided via the command line parameters.
+func BuildOpenstack(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	var config io.ReadCloser
+
+	// Should be loaded with --cloud-config /etc/kubernetes/kube_openstack_config from master node
+	if opts.CloudConfig != "" {
+		var err error
+		config, err = os.Open(opts.CloudConfig)
+		if err != nil {
+			klog.Fatalf("Couldn't open cloud provider configuration %s: %#v", opts.CloudConfig, err)
+		}
+		defer config.Close()
+	}
+
+	manager, err := CreateOpenstackManager(config, do, opts)
+	if err != nil {
+		klog.Fatalf("Failed to create openstack manager: %v", err)
+	}
+
+	provider, err := buildOpenstackCloudProvider(manager, rl)
+	if err != nil {
+		klog.Fatalf("Failed to create openstack cloud provider: %v", err)
+	}
+
+	if len(do.NodeGroupSpecs) == 0 {
+		klog.Fatalf("Must specify at least one node group with --nodes=<min>:<max>:<name>,...")
```

Comment:
This is also temporary? Isn't it possible for the auto scaler to fetch the node group list from Magnum dynamically?

Reply:
Yes this will be changed once there is an alternative (nodegroups).
I've added a TODO here in https://github.com/kubernetes/autoscaler/pull/1690/commits/b9216599ba159c385b50efd7a63b8a3b756ebb09.